src/main.rs

```diff
@@ -0,0 +1,81 @@
+use std::{error::Error, net::SocketAddr, sync::Arc};
+
+use quinn::{Endpoint, ServerConfig};
+
+/// Constructs a QUIC endpoint configured to listen for incoming connections on a certain address
+/// and port.
+///
+/// ## Returns
+///
+/// - a stream of incoming QUIC connections
+/// - server certificate serialized into DER format
+#[allow(unused)]
+pub fn make_server_endpoint(bind_addr: SocketAddr) -> Result<(Endpoint, Vec<u8>), Box<dyn Error>> {
+    let (server_config, server_cert) = configure_server()?;
+    let endpoint = Endpoint::server(server_config, bind_addr)?;
+    Ok((endpoint, server_cert))
+}
+
+/// Returns default server configuration along with its certificate.
+#[allow(clippy::field_reassign_with_default)] // https://github.com/rust-lang/rust-clippy/issues/6527
+fn configure_server() -> Result<(ServerConfig, Vec<u8>), Box<dyn Error>> {
+    let cert = rcgen::generate_simple_self_signed(vec!["localhost".into()]).unwrap();
+    let cert_der = cert.serialize_der().unwrap();
+    let priv_key = cert.serialize_private_key_der();
+    let priv_key = rustls::PrivateKey(priv_key);
+    let cert_chain = vec![rustls::Certificate(cert_der.clone())];
+
+    let mut server_config = ServerConfig::with_single_cert(cert_chain, priv_key)?;
+    Arc::get_mut(&mut server_config.transport)
+        .unwrap()
+        .max_concurrent_uni_streams(0_u8.into());
+
+    Ok((server_config, cert_der))
+}
+
+#[allow(unused)]
+pub const ALPN_QUIC_HTTP: &[&[u8]] = &[b"hq-29"];
+
+/// Runs a QUIC server bound to given address.
+async fn run_server(addr: SocketAddr) {
+    let (endpoint, _server_cert) = make_server_endpoint(addr).unwrap();
+    // accept a single connection
+    let incoming_conn = endpoint.accept().await;
+    if let Some(conn) = incoming_conn {
+        println!(
+            "[server] connection accepted: addr={}",
+            conn.remote_address()
+        );
+    }
+}
+
+/// Dummy certificate verifier that treats any certificate as valid.
```

Comment:
We do not want TLS or https for this exporter. All it does is expose metrics so no authentication is desired or needed in this use case.